packages/integrations/src/offline.ts

```diff
@@ -0,0 +1,169 @@
+        // cache if we are positively offline
+        if ('navigator' in this.global && 'onLine' in this.global.navigator && !this.global.navigator.onLine) {
+          this._cacheEvent(event)
+            .then((_event: Event): Promise<void> => this._enforceMaxEvents())
+            .catch(
+              (_error): void => {
+                logger.warn('could not cache event while offline');
+              },
+            );
+
+          // return null on success or failure, because being offline will still result in an error
+          return null;
+        }
+      }
+
+      return event;
+    });
+
+    // if online now, send any events stored in a previous offline session
+    if ('navigator' in this.global && 'onLine' in this.global.navigator && this.global.navigator.onLine) {
+      this._sendEvents().catch(() => {
+        logger.warn('could not send cached events');
+      });
+    }
+  }
+
+  /**
+   * cache an event to send later
+   * @param event an event
+   */
+  private async _cacheEvent(event: Event): Promise<Event> {
+    return this.offlineEventStore.setItem<Event>(uuid4(), event);
+  }
+
+  /**
+   * purge excess events if necessary
+   */
+  private async _enforceMaxEvents(): Promise<void> {
+    const events: Array<{ event: Event; cacheKey: string }> = [];
+
+    return this.offlineEventStore
+      .iterate<Event, void>(
+        (event: Event, cacheKey: string, _index: number): void => {
+          // aggregate events
+          events.push({ cacheKey, event });
+        },
+      )
+      .then(
+        (): Promise<void> =>
+          // this promise resolves when the iteration is finished
+          this._purgeEvents(
+            // purge all events past maxStoredEvents in reverse chronological order
+            events
+              .sort((a, b) => (b.event.timestamp || 0) - (a.event.timestamp || 0))
+              .slice(this.maxStoredEvents < events.length ? this.maxStoredEvents : events.length)
+              .map(event => event.cacheKey),
+          ),
+      )
+      .catch(
+        (_error): void => {
+          logger.warn('could not enforce max events');
+        },
+      );
+  }
+
+  /**
+   * purge event from cache
+   */
+  private async _purgeEvent(cacheKey: string): Promise<void> {
+    return this.offlineEventStore.removeItem(cacheKey);
+  }
+
+  /**
+   * purge events from cache
+   */
+  private async _purgeEvents(cacheKeys: string[]): Promise<void> {
+    // trail with .then to ensure the return type as void and not void|void[]
+    return Promise.all(cacheKeys.map(cacheKey => this._purgeEvent(cacheKey))).then();
+  }
+
+  /**
+   * send all events
+   */
+  private async _sendEvents(): Promise<void> {
+    return this.offlineEventStore.iterate<Event, void>(
+      (event: Event, cacheKey: string, _index: number): void => {
+        if (this.hub) {
+          const newEventId = this.hub.captureEvent(event);
+
+          if (newEventId) {
```

Comment:
The `Hub.captureEvent` method [always returns new event ID](https://github.com/getsentry/sentry-javascript/blob/2e761fe183a7e6d63fd438dc0aa9f071c2fff5ea/packages/hub/src/hub.ts#L210), so event will always be purged from event store (even that client failed to send it).

Reply:
Fixed in #2890